File: core/src/main/java/org/iumotionlabs/hefesto/feature/portinfo/service/ProcessEnrichmentService.java
package org.iumotionlabs.hefesto.feature.portinfo.service;

import org.iumotionlabs.hefesto.feature.portinfo.model.ProcessInfo;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.time.Instant;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Service for enriching process information with extended details.
 */
public final class ProcessEnrichmentService {

    private final PortInfoService.OperatingSystem os;

    public ProcessEnrichmentService() {
        this.os = PortInfoService.OperatingSystem.current();
    }

    /**
     * Gets extended process information for a PID.
     */
    public Optional<ProcessInfo> getProcessInfo(long pid) {
        return switch (os) {
            case LINUX -> getLinuxProcessInfo(pid);
            case MACOS -> getMacOsProcessInfo(pid);
            case WINDOWS -> getWindowsProcessInfo(pid);
            case UNKNOWN -> Optional.empty();
        };
    }

    private Optional<ProcessInfo> getLinuxProcessInfo(long pid) {
        try {
            // Use ps command to get process info
            Process process = new ProcessBuilder(
                "ps", "-p", String.valueOf(pid), "-o", "comm=,user=,rss=,vsz=,time=,nlwp=,lstart="
            ).redirectErrorStream(true).start();

            try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                String line = reader.readLine();
                if (line != null && !line.isBlank()) {
                    return parseLinuxPsOutput(pid, line);
                }
            }

            process.waitFor();
        } catch (Exception e) {
            // Fallback to basic info
        }

        return Optional.empty();
    }

    private Optional<ProcessInfo> parseLinuxPsOutput(long pid, String line) {
        // Format: name user rss vsz time nlwp lstart
        String[] parts = line.trim().split("\\s+", 7);
        if (parts.length >= 6) {
            try {
                String name = parts[0];
                String user = parts[1];
                long rssKb = Long.parseLong(parts[2]);
                long vszKb = Long.parseLong(parts[3]);
                long cpuTimeMs = parseCpuTime(parts[4]);
                int threadCount = Integer.parseInt(parts[5]);
                Instant startTime = parts.length > 6 ? parseLinuxLstart(parts[6]) : null;

                String cwd = getWorkingDirectory(pid);
                String cmdLine = getCommandLine(pid);

                return Optional.of(new ProcessInfo(
                    pid, name, cmdLine, user, cwd, rssKb, vszKb, cpuTimeMs, threadCount, startTime
                ));
            } catch (NumberFormatException e) {
                return Optional.empty();
            }
        }
        return Optional.empty();
    }

    private Optional<ProcessInfo> getMacOsProcessInfo(long pid) {
        try {
            // Use ps command to get process info
            Process process = new ProcessBuilder(
                "ps", "-p", String.valueOf(pid), "-o", "comm=,user=,rss=,vsz=,time=,lstart="
            ).redirectErrorStream(true).start();

            try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                String line = reader.readLine();
                if (line != null && !line.isBlank()) {
                    return parseMacOsPsOutput(pid, line);
                }
            }

            process.waitFor();
        } catch (Exception e) {
            // Fallback to basic info
        }

        return Optional.empty();
    }

    private Optional<ProcessInfo> parseMacOsPsOutput(long pid, String line) {
        // Format varies on macOS, using simpler parsing
        Pattern pattern = Pattern.compile("^(\\S+)\\s+(\\S+)\\s+(\\d+)\\s+(\\d+)\\s+(\\S+)\\s+(.*)$");
        Matcher m = pattern.matcher(line.trim());

        if (m.find()) {
            try {
                String name = m.group(1);
                String user = m.group(2);
                long rssKb = Long.parseLong(m.group(3));
                long vszKb = Long.parseLong(m.group(4));
                long cpuTimeMs = parseCpuTime(m.group(5));
                Instant startTime = parseMacOsLstart(m.group(6));

                String cwd = getWorkingDirectory(pid);
                String cmdLine = getCommandLine(pid);
                int threadCount = getThreadCount(pid);

                return Optional.of(new ProcessInfo(
                    pid, name, cmdLine, user, cwd, rssKb, vszKb, cpuTimeMs, threadCount, startTime
                ));
            } catch (NumberFormatException e) {
                return Optional.empty();
            }
        }
        return Optional.empty();
    }

    private Optional<ProcessInfo> getWindowsProcessInfo(long pid) {
        try {
            // Use WMIC to get process info
            Process process = new ProcessBuilder(
                "wmic", "process", "where", "ProcessId=" + pid, "get",
                "Name,CommandLine,WorkingSetSize,VirtualSize,UserModeTime,KernelModeTime,ThreadCount,CreationDate",
                "/format:csv"
            ).redirectErrorStream(true).start();

            try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                // Skip header
                reader.readLine();
                reader.readLine();
                String line = reader.readLine();
                if (line != null && !line.isBlank()) {
                    return parseWindowsWmicOutput(pid, line);
                }
            }

            process.waitFor();
        } catch (Exception e) {
            // Fallback
        }

        return Optional.empty();
    }

    private Optional<ProcessInfo> parseWindowsWmicOutput(long pid, String line) {
        // CSV format: Node,CommandLine,CreationDate,KernelModeTime,Name,ThreadCount,UserModeTime,VirtualSize,WorkingSetSize
        String[] parts = line.split(",");
        if (parts.length >= 9) {
            try {
                String cmdLine = parts[1];
                String name = parts[4];
                int threadCount = Integer.parseInt(parts[5].trim());
                long userModeTime = Long.parseLong(parts[6].trim());
                long kernelModeTime = Long.parseLong(parts[3].trim());
                long vszBytes = Long.parseLong(parts[7].trim());
                long rssBytes = Long.parseLong(parts[8].trim());

                long cpuTimeMs = (userModeTime + kernelModeTime) / 10000; // 100ns to ms
                long rssKb = rssBytes / 1024;
                long vszKb = vszBytes / 1024;

                return Optional.of(new ProcessInfo(
                    pid, name, cmdLine, "", "", rssKb, vszKb, cpuTimeMs, threadCount, null
                ));
            } catch (NumberFormatException e) {
                return Optional.empty();
            }
        }
        return Optional.empty();
    }

    private long parseCpuTime(String time) {
        // Format: HH:MM:SS or MM:SS
        String[] parts = time.split(":");
        try {
            if (parts.length == 3) {
                int hours = Integer.parseInt(parts[0]);
                int minutes = Integer.parseInt(parts[1]);
                int seconds = Integer.parseInt(parts[2].split("\\.")[0]);
                return (hours * 3600L + minutes * 60L + seconds) * 1000L;
            } else if (parts.length == 2) {
                int minutes = Integer.parseInt(parts[0]);
                int seconds = Integer.parseInt(parts[1].split("\\.")[0]);
                return (minutes * 60L + seconds) * 1000L;
            }
        } catch (NumberFormatException e) {
            // Ignore
        }
        return 0;
    }

    private Instant parseLinuxLstart(String lstart) {
        // Format: Day Mon DD HH:MM:SS YYYY
        // This is complex to parse, simplified version
        return null;
    }

    private Instant parseMacOsLstart(String lstart) {
        // Similar complex format
        return null;
    }

    private String getWorkingDirectory(long pid) {
        try {
            if (os == PortInfoService.OperatingSystem.LINUX) {
                Process process = new ProcessBuilder("readlink", "-f", "/proc/" + pid + "/cwd")
                    .redirectErrorStream(true).start();
                try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                    String line = reader.readLine();
                    return line != null ? line.trim() : "";
                }
            } else if (os == PortInfoService.OperatingSystem.MACOS) {
                Process process = new ProcessBuilder("lsof", "-p", String.valueOf(pid), "-Fn")
                    .redirectErrorStream(true).start();
                try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        if (line.startsWith("n") && line.contains("cwd")) {
                            return line.substring(1);
                        }
                    }
                }
            }
        } catch (Exception e) {
            // Ignore
        }
        return "";
    }

    private String getCommandLine(long pid) {
        try {
            if (os == PortInfoService.OperatingSystem.LINUX) {
                Process process = new ProcessBuilder("cat", "/proc/" + pid + "/cmdline")
                    .redirectErrorStream(true).start();
                try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                    String line = reader.readLine();
                    return line != null ? line.replace('\0', ' ').trim() : "";
                }
            } else if (os == PortInfoService.OperatingSystem.MACOS) {
                Process process = new ProcessBuilder("ps", "-p", String.valueOf(pid), "-o", "command=")
                    .redirectErrorStream(true).start();
                try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                    String line = reader.readLine();
                    return line != null ? line.trim() : "";
                }
            }
        } catch (Exception e) {
            // Ignore
        }
        return "";
    }

    private int getThreadCount(long pid) {
        try {
            if (os == PortInfoService.OperatingSystem.MACOS) {
                Process process = new ProcessBuilder("ps", "-M", "-p", String.valueOf(pid))
                    .redirectErrorStream(true).start();
                try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                    int count = 0;
                    while (reader.readLine() != null) {
                        count++;
                    }
                    return Math.max(0, count - 1); // Subtract header
                }
            }
        } catch (Exception e) {
            // Ignore
        }
        return 0;
    }
}

File: hefesto-core/src/feature/portinfo/service/process_enrichment_service.rs
use std::process::Command;

use hefesto_domain::portinfo::process_info::ProcessInfo;

/// Service for enriching process information with extended details.
///
/// Gathers additional process metadata (memory usage, CPU time, thread count,
/// working directory, full command line) using platform-specific commands.
pub struct ProcessEnrichmentService;

impl ProcessEnrichmentService {
    /// Creates a new `ProcessEnrichmentService`.
    pub fn new() -> Self {
        Self
    }

    /// Gets extended process information for a PID.
    pub fn get_process_info(&self, pid: u32) -> Option<ProcessInfo> {
        #[cfg(target_os = "linux")]
        {
            self.get_linux_process_info(pid)
        }

        #[cfg(target_os = "macos")]
        {
            self.get_macos_process_info(pid)
        }

        #[cfg(target_os = "windows")]
        {
            self.get_windows_process_info(pid)
        }

        #[cfg(not(any(target_os = "linux", target_os = "macos", target_os = "windows")))]
        {
            None
        }
    }

    // ── Linux ──────────────────────────────────────────────────────────────

    #[cfg(target_os = "linux")]
    fn get_linux_process_info(&self, pid: u32) -> Option<ProcessInfo> {
        let output = Command::new("ps")
            .args([
                "-p",
                &pid.to_string(),
                "-o",
                "comm=,user=,rss=,vsz=,time=,nlwp=",
            ])
            .output()
            .ok()?;

        let line = String::from_utf8_lossy(&output.stdout);
        let line = line.trim();
        if line.is_empty() {
            return None;
        }

        self.parse_linux_ps_output(pid, line)
    }

    #[cfg(target_os = "linux")]
    fn parse_linux_ps_output(&self, pid: u32, line: &str) -> Option<ProcessInfo> {
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() < 6 {
            return None;
        }

        let name = parts[0].to_string();
        let user = parts[1].to_string();
        let rss_kb: u64 = parts[2].parse().ok()?;
        let vsz_kb: u64 = parts[3].parse().ok()?;
        let cpu_time_ms = parse_cpu_time(parts[4]);
        let thread_count: u32 = parts[5].parse().unwrap_or(0);

        let cwd = self.get_working_directory(pid);
        let cmd_line = self.get_command_line(pid);

        Some(ProcessInfo {
            pid,
            name,
            command_line: cmd_line,
            user,
            working_directory: cwd,
            memory_rss_kb: rss_kb,
            memory_virtual_kb: vsz_kb,
            cpu_time_ms,
            thread_count,
            start_time: None,
        })
    }

    // ── macOS ──────────────────────────────────────────────────────────────

    #[cfg(target_os = "macos")]
    fn get_macos_process_info(&self, pid: u32) -> Option<ProcessInfo> {
        let output = Command::new("ps")
            .args(["-p", &pid.to_string(), "-o", "comm=,user=,rss=,vsz=,time="])
            .output()
            .ok()?;

        let line = String::from_utf8_lossy(&output.stdout);
        let line = line.trim();
        if line.is_empty() {
            return None;
        }

        self.parse_macos_ps_output(pid, line)
    }

    #[cfg(target_os = "macos")]
    fn parse_macos_ps_output(&self, pid: u32, line: &str) -> Option<ProcessInfo> {
        // Format: name user rss vsz time
        let re = regex::Regex::new(r"^(\S+)\s+(\S+)\s+(\d+)\s+(\d+)\s+(\S+)").ok()?;
        let captures = re.captures(line.trim())?;

        let name = captures.get(1)?.as_str().to_string();
        let user = captures.get(2)?.as_str().to_string();
        let rss_kb: u64 = captures.get(3)?.as_str().parse().ok()?;
        let vsz_kb: u64 = captures.get(4)?.as_str().parse().ok()?;
        let cpu_time_ms = parse_cpu_time(captures.get(5)?.as_str());

        let cwd = self.get_working_directory(pid);
        let cmd_line = self.get_command_line(pid);
        let thread_count = self.get_thread_count(pid);

        Some(ProcessInfo {
            pid,
            name,
            command_line: cmd_line,
            user,
            working_directory: cwd,
            memory_rss_kb: rss_kb,
            memory_virtual_kb: vsz_kb,
            cpu_time_ms,
            thread_count,
            start_time: None,
        })
    }

    // ── Windows ────────────────────────────────────────────────────────────

    #[cfg(target_os = "windows")]
    fn get_windows_process_info(&self, pid: u32) -> Option<ProcessInfo> {
        let output = Command::new("wmic")
            .args([
                "process",
                "where",
                &format!("ProcessId={pid}"),
                "get",
                "Name,CommandLine,WorkingSetSize,VirtualSize,UserModeTime,KernelModeTime,ThreadCount,CreationDate",
                "/format:csv",
            ])
            .output()
            .ok()?;

        let stdout = String::from_utf8_lossy(&output.stdout);
        let lines: Vec<&str> = stdout.lines().filter(|l| !l.trim().is_empty()).collect();

        // Skip header lines
        if lines.len() < 3 {
            return None;
        }

        self.parse_windows_wmic_output(pid, lines[2])
    }

    #[cfg(target_os = "windows")]
    fn parse_windows_wmic_output(&self, pid: u32, line: &str) -> Option<ProcessInfo> {
        let parts: Vec<&str> = line.split(',').collect();
        if parts.len() < 9 {
            return None;
        }

        let cmd_line = parts[1].to_string();
        let name = parts[4].to_string();
        let thread_count: u32 = parts[5].trim().parse().unwrap_or(0);
        let user_mode_time: u64 = parts[6].trim().parse().unwrap_or(0);
        let kernel_mode_time: u64 = parts[3].trim().parse().unwrap_or(0);
        let vsz_bytes: u64 = parts[7].trim().parse().unwrap_or(0);
        let rss_bytes: u64 = parts[8].trim().parse().unwrap_or(0);

        let cpu_time_ms = (user_mode_time + kernel_mode_time) / 10_000; // 100ns to ms
        let rss_kb = rss_bytes / 1024;
        let vsz_kb = vsz_bytes / 1024;

        Some(ProcessInfo {
            pid,
            name,
            command_line: cmd_line,
            user: String::new(),
            working_directory: String::new(),
            memory_rss_kb: rss_kb,
            memory_virtual_kb: vsz_kb,
            cpu_time_ms,
            thread_count,
            start_time: None,
        })
    }

    // ── Cross-platform helpers ─────────────────────────────────────────────

    fn get_working_directory(&self, pid: u32) -> String {
        #[cfg(target_os = "linux")]
        {
            Command::new("readlink")
                .args(["-f", &format!("/proc/{pid}/cwd")])
                .output()
                .ok()
                .and_then(|o| {
                    let s = String::from_utf8_lossy(&o.stdout).trim().to_string();
                    if s.is_empty() { None } else { Some(s) }
                })
                .unwrap_or_default()
        }

        #[cfg(target_os = "macos")]
        {
            Command::new("lsof")
                .args(["-p", &pid.to_string(), "-Fn"])
                .output()
                .ok()
                .and_then(|o| {
                    let stdout = String::from_utf8_lossy(&o.stdout);
                    for line in stdout.lines() {
                        if line.starts_with('n') && line.contains("cwd") {
                            return Some(line[1..].to_string());
                        }
                    }
                    None
                })
                .unwrap_or_default()
        }

        #[cfg(target_os = "windows")]
        {
            String::new()
        }

        #[cfg(not(any(target_os = "linux", target_os = "macos", target_os = "windows")))]
        {
            String::new()
        }
    }

    fn get_command_line(&self, pid: u32) -> String {
        #[cfg(target_os = "linux")]
        {
            Command::new("cat")
                .arg(format!("/proc/{pid}/cmdline"))
                .output()
                .ok()
                .map(|o| {
                    String::from_utf8_lossy(&o.stdout)
                        .replace('\0', " ")
                        .trim()
                        .to_string()
                })
                .unwrap_or_default()
        }

        #[cfg(target_os = "macos")]
        {
            Command::new("ps")
                .args(["-p", &pid.to_string(), "-o", "command="])
                .output()
                .ok()
                .map(|o| String::from_utf8_lossy(&o.stdout).trim().to_string())
                .unwrap_or_default()
        }

        #[cfg(target_os = "windows")]
        {
            String::new()
        }

        #[cfg(not(any(target_os = "linux", target_os = "macos", target_os = "windows")))]
        {
            String::new()
        }
    }

    #[cfg(target_os = "macos")]
    fn get_thread_count(&self, pid: u32) -> u32 {
        Command::new("ps")
            .args(["-M", "-p", &pid.to_string()])
            .output()
            .ok()
            .map(|o| {
                let count = String::from_utf8_lossy(&o.stdout).lines().count();
                // Subtract header line
                count.saturating_sub(1) as u32
            })
            .unwrap_or(0)
    }
}

impl Default for ProcessEnrichmentService {
    fn default() -> Self {
        Self::new()
    }
}

/// Parses a CPU time string in the format HH:MM:SS or MM:SS into milliseconds.
fn parse_cpu_time(time: &str) -> u64 {
    let parts: Vec<&str> = time.split(':').collect();
    match parts.len() {
        3 => {
            let hours: u64 = parts[0].parse().unwrap_or(0);
            let minutes: u64 = parts[1].parse().unwrap_or(0);
            let seconds: u64 = parts[2]
                .split('.')
                .next()
                .and_then(|s| s.parse().ok())
                .unwrap_or(0);
            (hours * 3600 + minutes * 60 + seconds) * 1000
        }
        2 => {
            let minutes: u64 = parts[0].parse().unwrap_or(0);
            let seconds: u64 = parts[1]
                .split('.')
                .next()
                .and_then(|s| s.parse().ok())
                .unwrap_or(0);
            (minutes * 60 + seconds) * 1000
        }
        _ => 0,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_cpu_time_hhmmss() {
        assert_eq!(parse_cpu_time("01:30:45"), (1 * 3600 + 30 * 60 + 45) * 1000);
    }

    #[test]
    fn test_parse_cpu_time_mmss() {
        assert_eq!(parse_cpu_time("05:30"), (5 * 60 + 30) * 1000);
    }

    #[test]
    fn test_parse_cpu_time_with_fractional() {
        assert_eq!(parse_cpu_time("01:30:45.123"), (1 * 3600 + 30 * 60 + 45) * 1000);
    }

    #[test]
    fn test_parse_cpu_time_zero() {
        assert_eq!(parse_cpu_time("00:00:00"), 0);
    }

    #[test]
    fn test_parse_cpu_time_invalid() {
        assert_eq!(parse_cpu_time("invalid"), 0);
    }
}
